### ISSUE TITLE
search: querying & sorting

### ISSUE THREAD

**Issue by kremalicious** [oceanprotocol/aquarius#181]
When doing a search for "ocean", all assets in a marketplace are returned. None of them has "ocean" in their base metadata. This makes it almost impossible to search for actual ocean-related data sets (ocean as in the watery thing)

### Expectation

When searching for "ocean" only assets are returned which actually have the word "ocean" in their `metadata.base`.

**Comment by aaitor** [comment]
The implementation of this feature using mongodb as backend is complicated. The `text search` in MongoDB doesn't support the search in a specific path of a nested document (`metadata.base`) in our case. 

It's an option to create a specific collection only for searching during the creation of the DDO. It means when you create a DDO it creates 2 collections:
* DDO - With the complete DDO document
* Metadata - Only with the metadata information, the search queries using that collection to avoid affect the search results

The problem of this approach is we duplicate the information in the DB and we need to handle the delete and update of information in 2 places, with the risk of data misalignment.

My suggestion is to support this functionality using the Elastic Search plugin in the short term.

**Comment by aaitor** [comment]
this issue depends on https://github.com/oceanprotocol/atlantic/issues/54

**Comment by jernejpregelj** [comment]
Testing ElasticSearch with Barge:

GET requets with text search work as intended:
`http://localhost:5000/api/v1/aquarius/assets/ddo/query?text=ocean` returns only ocean assets not all of them.

POST requests with text search give error:
`POST http://localhost:5000/api/v1/aquarius/assets/ddo/query`
`{"offset":25,"page":1,"query":{"text":["Test"]}}`
```
aquarius_1                   | 2019-06-18 08:50:53 c50b6aff3455 elasticsearch[10] WARNING GET http://elasticsearch:9200/aquarius/_search?q=Bussiness [status:400 request:0.006s]
aquarius_1                   | 2019-06-18 08:50:53 c50b6aff3455 flask.app[10] ERROR Exception on /api/v1/aquarius/assets/ddo/query [POST]
aquarius_1                   | Traceback (most recent call last):
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
aquarius_1                   |     response = self.full_dispatch_request()
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
aquarius_1                   |     rv = self.handle_user_exception(e)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
aquarius_1                   |     return cors_after_request(app.make_response(f(*args, **kwargs)))
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
aquarius_1                   |     reraise(exc_type, exc_value, tb)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
aquarius_1                   |     raise value
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
aquarius_1                   |     rv = self.dispatch_request()
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
aquarius_1                   |     return self.view_functions[rule.endpoint](**req.view_args)
aquarius_1                   |   File "/aquarius/aquarius/app/assets.py", line 626, in query_ddo
aquarius_1                   |     query_result = dao.query(search_model)
aquarius_1                   |   File "/aquarius/aquarius/app/dao.py", line 64, in query
aquarius_1                   |     query_result, count = self.oceandb.query(query)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/oceandb_elasticsearch_driver/plugin.py", line 163, in query
aquarius_1                   |     q=query_parsed[1]
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/utils.py", line 84, in _wrapped
aquarius_1                   |     return func(*args, params=params, **kwargs)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/__init__.py", line 819, in search
aquarius_1                   |     "GET", _make_path(index, "_search"), params=params, body=body
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/transport.py", line 353, in perform_request
aquarius_1                   |     timeout=timeout,
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/http_urllib3.py", line 251, in perform_request
aquarius_1                   |     self._raise_error(response.status, raw_data)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/base.py", line 178, in _raise_error
aquarius_1                   |     status_code, error_message, additional_info
aquarius_1                   | elasticsearch.exceptions.RequestError: RequestError(400, 'illegal_argument_exception', 'query malformed, empty clause found at [1:11]')
```

Querying empty database gives errors:
```
aquarius_1                   | 2019-06-18 08:10:35 c50b6aff3455 elasticsearch[10] WARNING GET http://elasticsearch:9200/_mapping/field/created [status:404 request:0.016s]
aquarius_1                   | 2019-06-18 08:10:35 c50b6aff3455 flask.app[10] ERROR Exception on /api/v1/aquarius/assets/ddo/query [POST]
aquarius_1                   | Traceback (most recent call last):
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
aquarius_1                   |     response = self.full_dispatch_request()
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
aquarius_1                   |     rv = self.handle_user_exception(e)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
aquarius_1                   |     return cors_after_request(app.make_response(f(*args, **kwargs)))
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
aquarius_1                   |     reraise(exc_type, exc_value, tb)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
aquarius_1                   |     raise value
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
aquarius_1                   |     rv = self.dispatch_request()
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
aquarius_1                   |     return self.view_functions[rule.endpoint](**req.view_args)
aquarius_1                   |   File "/aquarius/aquarius/app/assets.py", line 626, in query_ddo
aquarius_1                   |     query_result = dao.query(search_model)
aquarius_1                   |   File "/aquarius/aquarius/app/dao.py", line 64, in query
aquarius_1                   |     query_result, count = self.oceandb.query(query)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/oceandb_elasticsearch_driver/plugin.py", line 144, in query
aquarius_1                   |     self._mapping_to_sort(search_model.sort.keys())
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/oceandb_elasticsearch_driver/plugin.py", line 215, in _mapping_to_sort
aquarius_1                   |     if self.driver._es.indices.get_field_mapping(i)[self.driver._index]['mappings'] == {}:
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/utils.py", line 84, in _wrapped
aquarius_1                   |     return func(*args, params=params, **kwargs)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/indices.py", line 411, in get_field_mapping
aquarius_1                   |     params=params,
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/transport.py", line 353, in perform_request
aquarius_1                   |     timeout=timeout,
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/http_urllib3.py", line 251, in perform_request
aquarius_1                   |     self._raise_error(response.status, raw_data)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/base.py", line 178, in _raise_error
aquarius_1                   |     status_code, error_message, additional_info
aquarius_1                   | elasticsearch.exceptions.NotFoundError: NotFoundError(404, '{}')
```

```
aquarius_1                   | 2019-06-18 08:06:48 c50b6aff3455 elasticsearch[10] WARNING GET http://elasticsearch:9200/aquarius/_search [status:404 request:0.502s]
aquarius_1                   | 2019-06-18 08:06:48 c50b6aff3455 flask.app[10] ERROR Exception on /api/v1/aquarius/assets/ddo/query [GET]
aquarius_1                   | Traceback (most recent call last):
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
aquarius_1                   |     response = self.full_dispatch_request()
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
aquarius_1                   |     rv = self.handle_user_exception(e)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask_cors/extension.py", line 161, in wrapped_function
aquarius_1                   |     return cors_after_request(app.make_response(f(*args, **kwargs)))
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
aquarius_1                   |     reraise(exc_type, exc_value, tb)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
aquarius_1                   |     raise value
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
aquarius_1                   |     rv = self.dispatch_request()
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
aquarius_1                   |     return self.view_functions[rule.endpoint](**req.view_args)
aquarius_1                   |   File "/aquarius/aquarius/app/assets.py", line 571, in query_text
aquarius_1                   |     query_result = dao.query(search_model)
aquarius_1                   |   File "/aquarius/aquarius/app/dao.py", line 66, in query
aquarius_1                   |     query_result, count = self.oceandb.text_query(query)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/oceandb_elasticsearch_driver/plugin.py", line 192, in text_query
aquarius_1                   |     q=search_model.text
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/utils.py", line 84, in _wrapped
aquarius_1                   |     return func(*args, params=params, **kwargs)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/__init__.py", line 819, in search
aquarius_1                   |     "GET", _make_path(index, "_search"), params=params, body=body
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/transport.py", line 353, in perform_request
aquarius_1                   |     timeout=timeout,
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/http_urllib3.py", line 251, in perform_request
aquarius_1                   |     self._raise_error(response.status, raw_data)
aquarius_1                   |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/base.py", line 178, in _raise_error
aquarius_1                   |     status_code, error_message, additional_info
aquarius_1                   | elasticsearch.exceptions.NotFoundError: NotFoundError(404, 'index_not_found_exception', 'no such index', aquarius, index_or_alias)
```

**Comment by jernejpregelj** [comment]
Errors above were solved with v0.3.2 Aquarius
Pending:
Modify required filter `service.metadata.base.datePublished` to only `datePublished`
Modify required filter `service.metadata.base.dateCreated` to only `dateCreated`
